lib/value/src/kind/crud/get.rs

```diff
@@ -0,0 +1,474 @@
+                        let exact_len = largest_known_index.map_or(0, |x| x + 1);
+                        if exact_len >= len_required {
+                            // Make the index positive, then continue below.
+                            index += exact_len as isize;
+                        } else {
+                            // Out of bounds index.
+                            return Self::undefined();
+                        }
+                    }
+
+                    debug_assert!(index >= 0, "negative indices already handled");
+
+                    let index = index as usize;
+                    let mut kind = array
+                        .known()
+                        .get(&index.into())
+                        .cloned()
+                        .unwrap_or_else(|| array.unknown_kind());
+
+                    if !self.is_exact() {
+                        kind = kind.or_undefined();
+                    }
+                    kind.get_recursive(iter)
+                } else {
+                    Self::undefined()
+                }
+            }
+            Some(BorrowedSegment::CoalesceField(field)) => {
+                let field_kind = self.get_field(field);
+
+                // The remaining segments if this match succeeded.
+                #[allow(clippy::needless_collect)]
+                // Need to collect to prevent infinite recursive iterator type.
+                let match_iter = iter
+                    .clone()
+                    .skip_while(|segment| matches!(segment, BorrowedSegment::CoalesceField(_)))
+                    // Skip the `CoalesceEnd`, which always exists after `CoalesceFields`.
+                    .skip(1)
+                    .collect::<Vec<_>>();
+
+                // This is the resulting type, assuming the match succeeded.
+                let match_type = field_kind
+                    .clone()
+                    // This type is only valid if the match succeeded, which means this type wasn't undefined.
+                    .without_undefined()
+                    .get_recursive(match_iter.into_iter());
+
+                if !field_kind.contains_undefined() {
+                    // This coalesce field will always be defined, so skip the others.
+                    return match_type;
+                }
+
+                // The first field may or may not succeed. Try both.
+                self.get_recursive(iter).union(match_type)
+            }
+            Some(BorrowedSegment::Invalid) => {
+                // Value::get returns `None` in this case, which means the value is not defined.
+                Self::undefined()
+            }
+            None => self.clone(),
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use lookup::lookup_v2::{parse_path, OwnedPath};
+    use lookup::owned_path;
+    use std::collections::BTreeMap;
+
+    use super::*;
+    use crate::kind::Collection;
+    #[test]
+    #[allow(clippy::too_many_lines)]
+    fn test_at_path() {
```

Comment:
With the number of test cases, it feels like it'd be excellent to be able to do this programmatically with proptest.

Reply:
I plan on adding prop testing for typedefs in the future. I think these are useful in addition to that though.